recovery/elliptics_recovery/types/dc.py

```diff
@@ -234,18 +265,105 @@ def contains(addresses_with_backends, address, backend_id):
     return address_range
 
 
-def final_merge(ctx, results):
+def process_uncommitted(ctx, results):
+    '''
+    Removes uncommitted keys. If a key has any committed replicas, then this key is
+    appended to the file containing committed keys.
+    If uncommitted key's replica hasn't exceeded prepare timeout, then skip recovering of the key,
+    because the key is under writing and can be committed in the nearest future.
+    '''
+    if ctx.dry_run or ctx.safe:
+        return
+
+    node = elliptics.create_node(log_file=ctx.log_file,
+                                 log_level=int(ctx.log_level),
+                                 wait_timeout=ctx.wait_timeout,
+                                 flags=elliptics.config_flags.no_route_list,
+                                 net_thread_num=1,
+                                 io_thread_num=1,
+                                 remotes=ctx.remotes)
+    session = elliptics.newapi.Session(node)
+    session.trace_id = ctx.trace_id
+    session.exceptions_policy = elliptics.exceptions_policy.no_exceptions
+    session.set_filter(elliptics.filters.all_final)
+    session.ioflags |= elliptics.io_flags.cas_timestamp
+    session.timestamp = ctx.prepare_timeout
+
+    stats = ctx.stats['recover']
+    stats_cmd = ctx.stats['commands']
+
+    for r in results:
+        with open(r.filename, 'ab') as f:
+            for _, batch in groupby(enumerate(load_key_data(r.uncommitted_filename)),
+                                    key=lambda x: x[0] / ctx.batch_size):
+                batch = [item[1] for item in batch]
+                tasks = []
+                for key, key_infos in batch:
+                    for info in key_infos:
+                        if info.flags & elliptics.record_flags.uncommitted:
+                            tasks.append((key, info.group_id, info.size))
+
+                statuses = {} # (key, group_id) -> status
+                for attempt in range(ctx.attempts):
+                    if not tasks:
+                        break
+
+                    batch_sizes = defaultdict(int) # group_id -> batch_size
+                    for _, group_id, key_size in tasks:
+                        batch_sizes[group_id] += key_size
+
+                    timeouts = {group_id: max(60, batch_size / ctx.data_flow_rate)
```

Comment:
Why does it use `batch_size`? Each key is removed by a single command, why do you specify timeout as for full batch?